src/Workspaces/Core/Portable/Diagnostics/Extensions.cs

```diff
@@ -74,11 +75,20 @@ public static string GetAnalyzerId(this DiagnosticAnalyzer analyzer)
             return GetAssemblyQualifiedName(type);
         }
 
+        /// <summary>
+        /// Cache of a <see cref="Type"/> to its <see cref="Type.AssemblyQualifiedName"/>.  We cache this as the latter
+        /// computes and allocates expensively every time it is called.
+        /// </summary>
+        private static ImmutableSegmentedDictionary<Type, string> s_typeToAssemblyQualifiedName = ImmutableSegmentedDictionary<Type, string>.Empty;
```

Comment:
Would it be an issue that we are holding those Type objects

Reply:
Type instances are unique, so this is only per DiagnosticAnalyzer.  We don't incur an allocation asking for them in the future.

Reply:
Any benefit of using ImmutableSegmentedDictionary instead of framewsork ConcurrentDictionary?

Reply:
Yes, it prevents going on the large object heap.  And just tends to exhibit non-strange pits-of-fail (For example, enumerating a concurrentdictionary has no struct based enumeration).